entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -89,14 +97,49 @@ public void execute(EntityQueryRequest request, StreamObserver<ResultSetChunk> r
         .convertToEDSQuery(request, attrNameToEDSAttrMap.get(request.getEntityType()));
     Iterator<Document> documentIterator = entitiesCollection.search(
         DocStoreConverter.transform(tenantId.get(), query));
-    List<Entity> entities = convertDocsToEntities(documentIterator);
-
-    //Build result
-    //TODO : chunk response. For now sending everything in one chunk
-    responseObserver.onNext(convertEntitiesToResultSetChunk(
-        entities,
-        request.getSelectionList(),
-        attrNameToEDSAttrMap.get(request.getEntityType())));
+
+    ResultSetMetadata resultSetMetadata = ResultSetMetadata.newBuilder()
+        .addAllColumnMetadata(
+            () -> request.getSelectionList().stream().map(Expression::getColumnIdentifier)
+                .map(
+                    ColumnIdentifier::getColumnName)
+                .map(s -> ColumnMetadata.newBuilder().setColumnName(s).build()).iterator())
+        .build();
+    if (!documentIterator.hasNext()) {
+      ResultSetChunk.Builder resultBuilder = ResultSetChunk.newBuilder();
+      resultBuilder.setResultSetMetadata(resultSetMetadata);
+      resultBuilder.setIsLastChunk(true);
+      resultBuilder.setChunkId(0);
+      responseObserver.onNext(resultBuilder.build());
```

Comment:
We can call onCompleted() and return here right? Seems confusing to have code below this since it implies that we should keep going(and I know there is `while (documentIterator.hasNext())` below but still confusing.)

Reply:
yeah makes sense, let me update